### ISSUE TITLE
Function attachClient sometimes does not call Done() on ClientsWg.

### ISSUE THREAD

**Issue by NiklasCi** [mochi-mqtt/server#424]
I using this awesome mqtt broker to test mqtt client implementations. 
I have alot of small tests where i start the broker, publish a message and catch it with a hook.

I notice that sometimes the `server.Close` function indefinitely stuck because it waits for `ClientsWg` turning zero.
I checked where in the code `Add(1)` and `Done()` is called and added some logs above like this:

```go
// attachClient validates an incoming client connection and if viable, attaches the client
// to the server, performs session housekeeping, and reads incoming packets.
func (s *Server) attachClient(cl *Client, listener string) error {
	defer func() {
		s.Listeners.ClientsWg.Done()
		fmt.Println("DONE")
	}()
	fmt.Println("ADD")
	s.Listeners.ClientsWg.Add(1)
	
	// ...
}
```

I found out that sometimes `Add()` is called but not `Done()` which leads to stucking `server.Close()` call.
There is some race condition that needs to be investigated how that can happen.

Here my logs:
```
time=2024-08-01T14:13:02.464+02:00 level=INFO msg="added hook" hook=allow-all-auth
time=2024-08-01T14:13:02.464+02:00 level=INFO msg="added hook" hook=events-test
time=2024-08-01T14:13:02.464+02:00 level=INFO msg=initialised hook=events-test
time=2024-08-01T14:13:02.464+02:00 level=INFO msg="attached listener" id=04337e06-3f96-4f74-bb4d-286191bb5adb protocol=tcp address=127.0.0.1:1883
time=2024-08-01T14:13:02.464+02:00 level=INFO msg="mochi mqtt starting" version=2.6.5
time=2024-08-01T14:13:02.464+02:00 level=INFO msg="mochi mqtt server started"
ADD
ADD
DONE
time=2024-08-01T14:13:02.965+02:00 level=WARN msg="" listener=04337e06-3f96-4f74-bb4d-286191bb5adb error="read tcp 127.0.0.1:1883->127.0.0.1:35602: use of closed network connection"
ADD
DONE
time=2024-08-01T14:13:02.966+02:00 level=WARN msg="" listener=04337e06-3f96-4f74-bb4d-286191bb5adb error="read tcp 127.0.0.1:1883->127.0.0.1:35616: use of closed network connection"
ADD
DONE
time=2024-08-01T14:13:02.967+02:00 level=WARN msg="" listener=04337e06-3f96-4f74-bb4d-286191bb5adb error="read tcp 127.0.0.1:1883->127.0.0.1:35620: use of closed network connection"
ADD
DONE
time=2024-08-01T14:13:03.968+02:00 level=WARN msg="" listener=04337e06-3f96-4f74-bb4d-286191bb5adb error="read tcp 127.0.0.1:1883->127.0.0.1:35622: use of closed network connection"
ADD
DONE
time=2024-08-01T14:13:04.969+02:00 level=WARN msg="" listener=04337e06-3f96-4f74-bb4d-286191bb5adb error="read tcp 127.0.0.1:1883->127.0.0.1:35628: use of closed network connection"
ADD
DONE
time=2024-08-01T14:13:06.970+02:00 level=WARN msg="" listener=04337e06-3f96-4f74-bb4d-286191bb5adb error="read tcp 127.0.0.1:1883->127.0.0.1:35642: use of closed network connection"
panic: test timed out after 20s
```

**Comment by thedevop** [comment]
@NiklasCi, can you repeat this without any custom hooks, as hooks can block attachClient from exiting.

**Comment by NiklasCi** [comment]
Ok i created a simple test and i couldn't reproduce it. So it might be still a problem on my side. I don't know if there can be a race condition if the server is starting and the client connects before the server fully started?

**Comment by thedevop** [comment]
I'm not aware such race condition exist.

It's good that you're no longer seeing issue after removal of custom hooks. Closing this issue, feel free to re-open if you're able to reproduce it.

**Comment by NiklasCi** [comment]
> I'm not aware such race condition exist.
> 
> It's good that you're no longer seeing issue after removal of custom hooks. Closing this issue, feel free to re-open if you're able to reproduce it.

I still have the custom hooks, but i adapted the test and this behavior didn't occur yet. Maybe it will occcur again and the rewrite of the tests just made it less likely to occur.

**Comment by quarksus** [comment]
Reopening context: I hit this same underlying race independently while stress-testing an embedding project's mTLS work, and this time got a reliable, minimal reproduction (a real `mqtt.Server` + real TCP listener racing connect/disconnect churn against `Close()`, no application code involved) — triggers within ~150 iterations on an unpatched tree, either as `Close()` hanging (as originally reported here) or, under `-race`, a `sync: WaitGroup is reused before previous Wait has returned` panic.

Root cause: `TCP.Serve`'s accept loop spawns each connection's handler as an untracked goroutine, so `Listeners.ClientsWg.Add(1)` (inside `Server.attachClient`) has no ordering guarantee relative to a concurrent `Close()`'s `Wait()`.

Fix + regression tests in #540 (bundled with the related #488, since both share the same trigger condition and an end-to-end test needs both fixed to pass reliably). Feel free to reopen this issue if that's useful for tracking, or just close it out via #540 — up to you.
